### PR TITLE
[GPU] Add a new GPU plugin hint memory optimization mode : AGGRESSIVE_MEMORY_OPT

### PR DESCRIPTION
**Pull request by yeonbok**
### Details:
- Add a new GPU plugin hint memory optimization mode : AGGRESSIVE_MEMORY_OPT

### Tickets:
 - *ticket-id*

### CODE REVIEW

File: src/inference/include/openvino/runtime/intel_gpu/properties.hpp [hunk 1]
```diff
@@ -115,6 +115,14 @@ static constexpr Property<ov::hint::Priority> host_task_priority{"GPU_HOST_TASK_
...
  */
 static constexpr Property<int64_t> available_device_mem{"AVAILABLE_DEVICE_MEM_SIZE"};
+
+/**
+ * @brief This key specifies priority for optimization in GPU plugin for dynamic shapes either of memory or performance.
+ * - "memory" : Set memory management policy to optimize memory aggresively, which can affect peformmance drop.
+ * - "performance" : Prioritize performance always, which can increase memory usage.
+ * @ingroup ov_runtime_ocl_gpu_prop_cpp_api
+ */
+static constexpr Property<std::string> llm_opt_mode{"GPU_LLM_OPT_MODE"};
```
Comment (vladimir-paramuzov) [inline src/inference/include/openvino/runtime/intel_gpu/properties.hpp:125]: 1. It shouldn't have string type. Some enum is needed.
2. the behavior is not limited by llm, so the name is not fully relevant

Comment (vladimir-paramuzov) [reply src/inference/include/openvino/runtime/intel_gpu/properties.hpp:125]: Maybe
```
static constexpr Property<bool> enable_aggressive_memory_opt;
```

Comment (yeonbok) [reply src/inference/include/openvino/runtime/intel_gpu/properties.hpp]: Updated.

File: src/plugins/intel_gpu/src/graph/primitive_inst.cpp [hunk 2]
```diff
@@ -564,7 +564,8 @@ event::ptr primitive_inst::realloc_if_needed() {
 
 
     // If we allocated too large memory, reclaim the memory.
-    if (updated_layout.get_buffer_size().count() * 10 < _max_output_layout_count) {
+    if (get_network().get_config().get_property(ov::intel_gpu::hint::aggressive_memory_opt)
```
Comment (vladimir-paramuzov) [inline src/plugins/intel_gpu/src/graph/primitive_inst.cpp:567]: Please save this flag somewhere. Query of the property on each iteration may be expensive.

Comment (yeonbok) [reply src/plugins/intel_gpu/src/graph/primitive_inst.cpp]: Applied

File: src/inference/include/openvino/runtime/intel_gpu/properties.hpp [hunk 3]
```diff
@@ -115,6 +115,13 @@ static constexpr Property<ov::hint::Priority> host_task_priority{"GPU_HOST_TASK_
  * @ingroup ov_runtime_ocl_gpu_prop_cpp_api
  */
 static constexpr Property<int64_t> available_device_mem{"AVAILABLE_DEVICE_MEM_SIZE"};
+
+/**
+ * @brief This key enables GPU plugin to optimize memory usage on dynamic shape execution.
+ * Note that setting this option true can cause peformance drop.
+ * @ingroup ov_runtime_ocl_gpu_prop_cpp_api
+ */
+static constexpr Property<bool> aggressive_memory_opt{"AGGRESSIVE_MEMORY_OPT"};
```
Comment (sshlyapn) [inline src/inference/include/openvino/runtime/intel_gpu/properties.hpp:124]: Shouldn't we add corresponding Python API property as well?

Comment (yeonbok) [reply src/inference/include/openvino/runtime/intel_gpu/properties.hpp:124]: I tested in python app and it worked. Should we do something else?

Comment (sshlyapn) [reply src/inference/include/openvino/runtime/intel_gpu/properties.hpp:124]: It provides more convenient python namespace for hint, something like intel_gpu_hint.aggressive_memory_opt, but if functionally it works, I think we may add this later

**Review by vladimir-paramuzov — APPROVED** [review]

File: src/inference/include/openvino/runtime/intel_gpu/properties.hpp (hunk 3 above)
Comment (yeonbok) [reply src/inference/include/openvino/runtime/intel_gpu/properties.hpp:124]: I see! Lets add it in the next PR. Thanks!

**Pull request closed without merging**